### ISSUE TITLE
[2.1.0] HTML minification problem with php tag with a comment and no space at the end

### ISSUE THREAD

**Issue by hostep** [magento/magento2#5316]
## Steps to reproduce
1. Install Magento 2.1.0 using composer
2. Create a theme and a template file which contains this:

``` php
<?php //if ($showDescription):?>
    <div class="product description product-item-description">
        <?php /* @escapeNotVerified */ echo $_helper->productAttribute($_product, $_product->getShortDescription(), 'short_description') ?>
    </div>
<?php //endif; ?>
```
1. Turn on html minification in the backend
## Expected result
1. Page is rendered properly without errors
## Actual result
1. Error message appears: `Parse error: syntax error, unexpected '<'`.

The minified php/html file was turned into:

```
<?php <div class="product description product-item-description"><?php /* @escapeNotVerified */ echo $_helper->productAttribute($_product, $_product->getShortDescription(), 'short_description') ?></div><?php ?>
```
## Temporary solution

Add a space at the end and it works again, like this:

``` php
<?php //if ($showDescription): ?>
```

**Comment by hostep** [comment]
Any updates on this?

@guz-anton, @okorshenko, @piotrekkaminski: since you guys were involved in another html minification issue, I'm including you in this conversation.

**Comment by anton-huz** [comment]
Hi @hostep ,
We are going to find solution under MAGETWO-55356.

**Comment by m0zziter** [comment]
Experienced a similar issue when I installed https://github.com/algolia/algoliasearch-magento-2. 

Except the problem line in my case was this line: https://github.com/algolia/algoliasearch-magento-2/blob/master/view/frontend/templates/internals/commonjs.phtml#L41

When html minification is turned on that line looks like
`if ($path != '') { $path .= ' }`

But I'd expect it to look like 
`if ($path != '') { $path .= ' /// '; }`

Here's a gist of the entire file: https://gist.github.com/m0zziter/b7c2f00afd3684e0642060d19617ab2f

**Comment by peec** [comment]
Same problem. 2.1.1.

**Comment by jmtakahashi** [comment]
This issue hasn't been closed yet and it doesn't seem like there has been a fix to the Minify HTML  problem.  @guz-anton @okorshenko @piotrekkaminski is there any update on if this issue has been fixed any of the 2.1 release?  Thanks.

**Comment by ooples** [comment]
I'm using 2.1.6 and I'm having the same issues with html minimication

**Comment by chenshanmin** [comment]
Does anyone has a solution?

**Comment by hostep** [comment]
Yes: add a space at the end, like I mention in my initial post under 'temporary solution' ;)

**Comment by chenshanmin** [comment]
@hostep maybe that is the best way now,but my project is a secondary development, how can I find all of the End tag :'?>' and add a space...  minification library is vendor/magento/framework/View/Template/Html/Minifier.php
right? can I rewrite it?

**Comment by pantaoran** [comment]
@hostep You have to be joking with your temporary solution. I just ran a regex to find all offending occurrencies in core phtml templates. It found 1109 cases within app/code/Magento.

@chenshanmin I guess it would be possible to try and do this with regular expressions search & replace.
I used the regex `[^\s]\?>` to search for any non-whitespace characters followed by a php closing tag. You could modify that to insert a space, for example with "sed" on bash.

**Comment by hostep** [comment]
@pantaoran: it's only for php tags with php code which is commented out in phtml files. I haven't searched the Magento core code if there are offending files like that, in our case it was a file in our custom theme.
I don't think your regex searches for commented out code?

Although I agree that all phtml files which contain `{something}?>` should be replaced with `{something} ?>` (including a space) to have everything consistent.

**Comment by pantaoran** [comment]
Apologies, I should have read it more carefully. Yes, my regex searches for any closing tags and does not consider the presence of comments, and it is therefore irrelevant here.

The following regex could work instead: `\/\/.*[^\s]\?>`
With this I found various occurrencies in third party modules and themes, but none in Magento core.

**Comment by chenshanmin** [comment]
I repeated all php tags  Ha ha ha

**Comment by pniel-cohen** [comment]
I used @pantaoran 's regex to find & fix a few template files, but the minified HTML is still broken.
Any other idea on how to debug this issue?

**Comment by franckgarnier21** [comment]
Any fix in Magento 2.2RC ?

**Comment by magento-engcom-team** [comment]
@hostep, thank you for your report.
We've created internal ticket(s) MAGETWO-83074 to track progress on the issue.

**Comment by srenon** [comment]
I'm having a similar issue with version 2.1.9 and 2.2.2.

Magento mode: default 

phtml code (with space at the end)
```
        </form>
    </div>
    <?php //echo $this->getLayout()->createBlock('Magento\Cms\Block\Block')->setBlockId('footer-links-block-4')->toHtml(); ?>
</div>
```
Error
>Parse error: syntax error, unexpected '/'

Minify HTML in var/view_preprocessed/pub/...

```
 </form></div><?php </div>
```

But if I changed the code to
```
<?php //echo 'test comment'; ?>
```

NO error

Minify HTML in var/view_preprocessed/pub/...
```
</form></div><?php ?></div>
```

My workaround was to check the comment to ``/* ... */``
```
<?php /* echo $this->getLayout()->createBlock('Magento\Cms\Block\Block')->setBlockId('footer-links-block-4')->toHtml(); */ ?>
```

**Comment by gabs77** [comment]
When we have /// in the data:image/gif;base64, the minify html remove all the datas in the image.

Ex: 
data:image/gif;base64,R0lGODlhEAAQAPQAAP///wAAAPDw8IqKiuDg4EZGRnp6egAAAFhYWCQkJKysrL6+vhQUFJycnAQEBDY2NmhoaAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACH/C05FVFNDQVBFMi4wAwEAAAAh/hpDcmVhdGVkIHdpdGggYWpheGxvYWQuaW5mbwAh+QQJCgAAACwAAAAAEAAQAAAFdyAgAgIJIeWoAkRCCMdBkKtIHIngyMKsErPBYbADpkSCwhDmQCBethRB6Vj4kFCkQPG4IlWDgrNRIwnO4UKBXDufzQvDMaoSDBgFb886MiQadgNABAokfCwzBA8LCg0Egl8jAggGAA1kBIA1BAYzlyILczULC2UhACH5BAkKAAAALAAAAAAQABAAAAV2ICACAmlAZTmOREEIyUEQjLKKxPHADhEvqxlgcGgkGI1DYSVAIAWMx+lwSKkICJ0QsHi9RgKBwnVTiRQQgwF4I4UFDQQEwi6/3YSGWRRmjhEETAJfIgMFCnAKM0KDV4EEEAQLiF18TAYNXDaSe3x6mjidN1s3IQAh+QQJCgAAACwAAAAAEAAQAAAFeCAgAgLZDGU5jgRECEUiCI+yioSDwDJyLKsXoHFQxBSHAoAAFBhqtMJg8DgQBgfrEsJAEAg4YhZIEiwgKtHiMBgtpg3wbUZXGO7kOb1MUKRFMysCChAoggJCIg0GC2aNe4gqQldfL4l/Ag1AXySJgn5LcoE3QXI3IQAh+QQJCgAAACwAAAAAEAAQAAAFdiAgAgLZNGU5joQhCEjxIssqEo8bC9BRjy9Ag7GILQ4QEoE0gBAEBcOpcBA0DoxSK/e8LRIHn+i1cK0IyKdg0VAoljYIg+GgnRrwVS/8IAkICyosBIQpBAMoKy9dImxPhS+GKkFrkX+TigtLlIyKXUF+NjagNiEAIfkECQoAAAAsAAAAABAAEAAABWwgIAICaRhlOY4EIgjH8R7LKhKHGwsMvb4AAy3WODBIBBKCsYA9TjuhDNDKEVSERezQEL0WrhXucRUQGuik7bFlngzqVW9LMl9XWvLdjFaJtDFqZ1cEZUB0dUgvL3dgP4WJZn4jkomWNpSTIyEAIfkECQoAAAAsAAAAABAAEAAABX4gIAICuSxlOY6CIgiD8RrEKgqGOwxwUrMlAoSwIzAGpJpgoSDAGifDY5kopBYDlEpAQBwevxfBtRIUGi8xwWkDNBCIwmC9Vq0aiQQDQuK+VgQPDXV9hCJjBwcFYU5pLwwHXQcMKSmNLQcIAExlbH8JBwttaX0ABAcNbWVbKyEAIfkECQoAAAAsAAAAABAAEAAABXkgIAICSRBlOY7CIghN8zbEKsKoIjdFzZaEgUBHKChMJtRwcWpAWoWnifm6ESAMhO8lQK0EEAV3rFopIBCEcGwDKAqPh4HUrY4ICHH1dSoTFgcHUiZjBhAJB2AHDykpKAwHAwdzf19KkASIPl9cDgcnDkdtNwiMJCshACH5BAkKAAAALAAAAAAQABAAAAV3ICACAkkQZTmOAiosiyAoxCq+KPxCNVsSMRgBsiClWrLTSWFoIQZHl6pleBh6suxKMIhlvzbAwkBWfFWrBQTxNLq2RG2yhSUkDs2b63AYDAoJXAcFRwADeAkJDX0AQCsEfAQMDAIPBz0rCgcxky0JRWE1AmwpKyEAIfkECQoAAAAsAAAAABAAEAAABXkgIAICKZzkqJ4nQZxLqZKv4NqNLKK2/Q4Ek4lFXChsg5ypJjs1II3gEDUSRInEGYAw6B6zM4JhrDAtEosVkLUtHA7RHaHAGJQEjsODcEg0FBAFVgkQJQ1pAwcDDw8KcFtSInwJAowCCA6RIwqZAgkPNgVpWndjdyohACH5BAkKAAAALAAAAAAQABAAAAV5ICACAimc5KieLEuUKvm2xAKLqDCfC2GaO9eL0LABWTiBYmA06W6kHgvCqEJiAIJiu3gcvgUsscHUERm+kaCxyxa+zRPk0SgJEgfIvbAdIAQLCAYlCj4DBw0IBQsMCjIqBAcPAooCBg9pKgsJLwUFOhCZKyQDA3YqIQAh+QQJCgAAACwAAAAAEAAQAAAFdSAgAgIpnOSonmxbqiThCrJKEHFbo8JxDDOZYFFb+A41E4H4OhkOipXwBElYITDAckFEOBgMQ3arkMkUBdxIUGZpEb7kaQBRlASPg0FQQHAbEEMGDSVEAA1QBhAED1E0NgwFAooCDWljaQIQCE5qMHcNhCkjIQAh+QQJCgAAACwAAAAAEAAQAAAFeSAgAgIpnOSoLgxxvqgKLEcCC65KEAByKK8cSpA4DAiHQ/DkKhGKh4ZCtCyZGo6F6iYYPAqFgYy02xkSaLEMV34tELyRYNEsCQyHlvWkGCzsPgMCEAY7Cg04Uk48LAsDhRA8MVQPEF0GAgqYYwSRlycNcWskCkApIyEAOwAAAAAAAAAAAA==

Become : data:image/gif;base64,R0lGODlhEAAQAPQAAP

We can exclude this case in the minify

**Comment by ghost** [comment]
Hi folks, any update on this? will this be fixed in 2.3 ?

**Comment by RahulKachhadia** [comment]
I am working on this at #dmcdindia

**Comment by magento-engcom-team** [comment]
@rahul-kachhadiya thank you for joining. Please accept team invitation [here](https://github.com/orgs/magento/invitation) and self-assign the issue.

**Comment by lisovyievhenii** [comment]
#kiev-cd

**Comment by magento-engcom-team** [comment]
Hi @hostep. Thank you for your report. 
The issue has been fixed in magento/magento2#16332 by @lisovyievhenii in 2.3-develop branch
Related commit(s):
 - [544c7b66b9ae43279cf6da5e8f5420228e24e0ff](https://github.com/magento/magento2/commit/544c7b66b9ae43279cf6da5e8f5420228e24e0ff)
 - [9804cb10e0212a19a02135de072f28b78d18a2a4](https://github.com/magento/magento2/commit/9804cb10e0212a19a02135de072f28b78d18a2a4)
 - [721fe0f24e662bbe15e2e7fae0dc3b9305ccdbc7](https://github.com/magento/magento2/commit/721fe0f24e662bbe15e2e7fae0dc3b9305ccdbc7)

The fix will be available with the upcoming 2.3.0 release.

**Comment by sidolov** [comment]
Hi @hostep. Thank you for your report. 
The issue has been fixed in magento/magento2#16917 by @ronak2ram in 2.1-develop branch
Related commit(s):
 - [28ef78dcea296f1e47ff119ccd4926df6af5ded0](https://github.com/magento/magento2/commit/28ef78dcea296f1e47ff119ccd4926df6af5ded0)
 - [433a9c36c86afb942aac1b589a74a13255ebdacb](https://github.com/magento/magento2/commit/433a9c36c86afb942aac1b589a74a13255ebdacb)

The fix will be available with the upcoming 2.1.15 release.

**Comment by VladimirZaets** [comment]
Hi @hostep. Thank you for your report. 
The issue has been fixed in magento/magento2#16916 by @ronak2ram in 2.2-develop branch
Related commit(s):
 - [bf86012edc5c790762aa927bbda664dc2ad64069](https://github.com/magento/magento2/commit/bf86012edc5c790762aa927bbda664dc2ad64069)
 - [7f0bce469f70c55fc9beb044e3c23a6146b262c7](https://github.com/magento/magento2/commit/7f0bce469f70c55fc9beb044e3c23a6146b262c7)

The fix will be available with the upcoming 2.2.6 release.

**Comment by VivekShingala** [comment]
Worked for me. Thanks @hostep
